src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp

```diff
@@ -3193,39 +3235,37 @@ CTranslatorExprToDXL::PdxlnQuantifiedSubplan(
 		pexprInner, NULL /*colref_array*/, pdrgpdsBaseTables,
 		pulNonGatherMotions, pfDML, false /*fRemap*/, false /*fRoot*/);
 
-	// find required column from inner child
-	CColRefSet *pcrInner = GPOS_NEW(m_mp) CColRefSet(m_mp);
-	pcrInner->Include((*pdrgpcrInner)[0]);
+	// In fact, the inner (right) part of CorrelatedNLJoin is transformed
+	// into a subquery. The scalar part of CorrelatedNLJoin is transformed
+	// into a Test Expression of the subquery.
+	// At execution, the subquery returns a tuple, and Test Expression checks
+	// the condition for returning the tuple - it acts as a filter.
+	// We do not need all the output columns of the subquery, so we need to
+	// restrict the projection of the inner node.
+	// We need the "Reqd Inner Col" of CorrelatedNLJoin (pdrgpcrInner[0])
+	// and the columns from the scalar that appear in the subquery (inner part).
+	// Since the "Reqd Inner Col" always appears in the scalar (this is true
+	// for ANY/ALL subqueries), we need to define the used columns in the Test
+	// Expression that are output in the inner node.
+	CColRefSet *pcrsInner =
+		GPOS_NEW(m_mp) CColRefSet(m_mp, *pexprInner->DeriveOutputColumns());
+	CColRefSet *pcrsScalarUsed = pexprScalar->DeriveUsedColumns();
+	pcrsInner->Intersection(pcrsScalarUsed);
+	GPOS_ASSERT(pcrsInner->Size() > 0);
```

Comment:
It seems that pcrsInner contains the same data as pcrsUsed in the deleted code. Why is `if` replaced with `GPOS_ASSERT`? Why is size=0 now considered an error, but this case was considered a normal situation before?